### ISSUE TITLE
Elastic Agent on GKE cluster - 1/9 agents cannot connect to Elastic Cloud - no more kubernetes.node.status.ready reported for none of them

### ISSUE THREAD

**Issue by Danouchka** [elastic/elastic-agent#2718]
Hello 

**Context**
I have deployed elastic-agent 8.7.1 on a GKE cluster (1.25.8-gke.500) made of 9 nodes
The agents are running as per the following gke nodes

```
elastic-agent-77mkw                                           3/3     Running   0              2d9h      gke-sa-da-gke-cluster-default-pool-a25db8ea-7cad   <none>           <none>
elastic-agent-996k6                                           3/3     Running   1 (2d5h ago)   2d9h     gke-sa-da-gke-cluster-default-pool-a25db8ea-2w0c   <none>           <none>
elastic-agent-dnbt5                                           3/3     Running   0              2d9h     gke-sa-da-gke-cluster-default-pool-a25db8ea-4pan   <none>           <none>
elastic-agent-f8z7t                                           3/3     Running   0              2d9h     gke-sa-da-gke-cluster-default-pool-a25db8ea-mmoi   <none>           <none>
elastic-agent-m2tq5                                           3/3     Running   0              2d9h     gke-sa-da-gke-cluster-default-pool-a25db8ea-699g   <none>           <none>
elastic-agent-rrn48                                           3/3     Running   0              2d9h     gke-sa-da-gke-cluster-default-pool-a25db8ea-n4ta   <none>           <none>
elastic-agent-svlrm                                           3/3     Running   0              2d9h     gke-sa-da-gke-cluster-default-pool-a25db8ea-exvu   <none>           <none>
elastic-agent-t7v72                                           3/3     Running   1 (2d6h ago)   2d9h     gke-sa-da-gke-cluster-default-pool-a25db8ea-rqhq   <none>           <none>
elastic-agent-v85ds                                           3/3     Running   0              2d9h     gke-sa-da-gke-cluster-default-pool-a25db8ea-xwqo   <none>           <none>
```

All nodes seems to be running fine 
<img width="1441" alt="Capture d’écran 2023-05-19 à 19 09 57" src="https://github.com/elastic/elastic-agent/assets/48563314/fdc69f75-00cd-45c9-ac11-fb8e78535270">

**Issue 1**
elastic-agent-dnbt5  is offline and no more reporting any data to Elastic Cloud 
 
<img width="1364" alt="Capture d’écran 2023-05-19 à 18 59 51" src="https://github.com/elastic/elastic-agent/assets/48563314/f451ae4a-3598-4ae9-b9be-f4c46e048d7b">

<img width="1878" alt="Capture d’écran 2023-05-19 à 18 59 23" src="https://github.com/elastic/elastic-agent/assets/48563314/c7046ecc-4f75-43e9-851a-5675daa40cd9">

Regarding on google cloud console, in log explorer, we can can see that actually, for an unknow reason, the elastic agent cannot connect anymore to elastic cloud 
```
{
  "insertId": "42fjms0fy84vb7gl",
  "jsonPayload": {
    "log.origin": {
      "file.line": 369,
      "file.name": "management/managerV2.go"
    },
    "log.logger": "centralmgmt.V2-manager",
    "service.name": "metricbeat",
    "log.level": "error",
    "ecs.version": "1.6.0",
    "message": "elastic-agent-client error: rpc error: code = Unavailable desc = connection error: desc = \"transport: authentication handshake failed: context deadline exceeded\"",
    "component": {
      "binary": "metricbeat",
      "id": "system/metrics-default",
      "type": "system/metrics",
      "dataset": "elastic_agent.metricbeat"
    },
    "log": {
      "source": "system/metrics-default"
    },
    "@timestamp": "2023-05-19T17:04:24.423Z"
  },
  "resource": {
    "type": "k8s_container",
    "labels": {
      "container_name": "elastic-agent",
      "pod_name": "elastic-agent-dnbt5",
      "namespace_name": "kube-system"
    }
  },
  "timestamp": "2023-05-19T17:04:24.585571737Z",
  "severity": "ERROR",
  "labels": {
    "compute.googleapis.com/resource_name": "gke-sa-da-gke-cluster-default-pool-a25db8ea-4pan",
    "k8s-pod/pod-template-generation": "1",
    "k8s-pod/app": "elastic-agent",
    "k8s-pod/controller-revision-hash": "55c5889b7f"
  },
  "logName": "projects/myproject/logs/stderr",
  "receiveTimestamp": "2023-05-19T17:04:27.399219324Z"
}
```
**Issue 2**
For strange reasons, it seems that following this issue, kubernetes.status.node.ready is no more reported for all elastic-agents (confirmed in metrics-* dataview) despite the 8 others are reporting their data properly except this one . 
Instead of seeing 11 nodes here, we see only 2 (from another cluster) in the top left visualization 
<img width="2075" alt="Capture d’écran 2023-05-19 à 19 01 25" src="https://github.com/elastic/elastic-agent/assets/48563314/3516a288-1313-47f2-9594-e75a8a002079">

Are these known issues ?
Thanks for fixing this

**Comment by Danouchka** [comment]
cc @jlind23 @pierrehilbert @michalpristas

**Comment by Danouchka** [comment]
I also confirm that elastic-agent-dnbt5 had no over cpu or memory consumption at the moment where it lost contact with elastic cloud 
<img width="2370" alt="Capture d’écran 2023-05-19 à 19 28 10" src="https://github.com/elastic/elastic-agent/assets/48563314/a91d9272-9fc0-4712-bb01-a7156d94f8e6">

**Comment by jlind23** [comment]
@MichaelKatsoulis @ChrsMark Can we please get your help here?

**Comment by ChrsMark** [comment]
cc: @elastic/obs-cloudnative-monitoring @bturquet @tommyers-elastic

**Comment by tommyers-elastic** [comment]
@jlind23 do we ever see this timeout error in non k8s agent deployments? 

on issue 2, @Danouchka did the nodes only stop reporting `kubernetes.status.node.ready` after that single node dropped? would we be able to take a look at the metric documents themselves (i.e. in 'Discover') for the nodes before and after the issue ?

**Comment by Danouchka** [comment]
Hi 
access to the logs & metrics granted to @tommyers-elastic

**Comment by Danouchka** [comment]
@tommyers-elastic said maybe related to that one https://github.com/elastic/sdh-beats/issues/3363

**Comment by gizas** [comment]
@Danouchka just was checking your cluster status:

![Screenshot 2023-05-23 at 10 43 46 AM](https://github.com/elastic/elastic-agent/assets/1685415/a234a405-91bb-4fe1-a63b-461a6a6ec7f4)

I can see now that 8 nodes are reporting. So is any of the agents still in "stopped" state?
Can you also add here some logs from the non-working agent to see if we can find something inside?

@elastic/elastic-agent-control-plane team I guess [this is](https://github.com/elastic/beats/blob/main/x-pack/libbeat/management/managerV2.go#L412-L415) the code that logs the messages above and I see [this reference](https://github.com/elastic/beats/blob/main/x-pack/libbeat/management/managerV2.go#L431-L434). Struggling to understand what corner case we hit here. Do you think is somehow related?

**Comment by jlind23** [comment]
Maybe @cmacknz @pierrehilbert can help us out here.

**Comment by Danouchka** [comment]
elastic-agent-77mkw                                           3/3     Running   0               6d       gke-sa-da-gke-cluster-default-pool-a25db8ea-7cad   <none>           <none>
elastic-agent-996k6                                           3/3     Running   1 (5d20h ago)   6d       gke-sa-da-gke-cluster-default-pool-a25db8ea-2w0c   <none>           <none>
elastic-agent-dnbt5                                           3/3     Running   1 (3d13h ago)   6d       gke-sa-da-gke-cluster-default-pool-a25db8ea-4pan   <none>           <none>
elastic-agent-f8z7t                                           3/3     Running   0               6d       gke-sa-da-gke-cluster-default-pool-a25db8ea-mmoi   <none>           <none>
elastic-agent-m2tq5                                           3/3     Running   0               6d       gke-sa-da-gke-cluster-default-pool-a25db8ea-699g   <none>           <none>
elastic-agent-rrn48                                           3/3     Running   0               6d       gke-sa-da-gke-cluster-default-pool-a25db8ea-n4ta   <none>           <none>
elastic-agent-svlrm                                           3/3     Running   1 (36h ago)     6d       gke-sa-da-gke-cluster-default-pool-a25db8ea-exvu   <none>           <none>
elastic-agent-t7v72                                           3/3     Running   1 (5d20h ago)   6d       gke-sa-da-gke-cluster-default-pool-a25db8ea-rqhq   <none>           <none>
elastic-agent-v85ds                                           3/3     Running   0               6d       gke-sa-da-gke-cluster-default-pool-a25db8ea-xwqo   <none>           <none>

**Comment by Danouchka** [comment]
I think did not had the issue in 8.6.2

**Comment by Danouchka** [comment]
The agent went offline 8 hours ago and there's a strong correlation that if one agent goes offline, metrics below are no more reported 
<img width="1692" alt="Capture d’écran 2023-05-23 à 10 01 17" src="https://github.com/elastic/elastic-agent/assets/48563314/cb6e8f71-b882-4950-bc6c-69875ea2fa8b">

**Comment by gizas** [comment]
Can you please run the command:
```bash
kubectl get leases -n kube-system
```
(To see which agent is the leader at the moment! )

**Comment by Danouchka** [comment]
NAME                                      HOLDER                                                                       AGE
addon-manager                             gke-94c9cea96359465dabd0-ce53-3e36-vm_b0d29db8-389f-4511-b3bb-881348778e8c   390d
cloud-controller-manager                  gke-94c9cea96359465dabd0-ce53-3e36-vm_c5f6fb3d-3f7f-46f0-a960-5150f71a8aa8   6d1h
cloud-provider-extraction-migration-pt1   gke-94c9cea96359465dabd0-ce53-3e36-vm_c5f6fb3d-3f7f-46f0-a960-5150f71a8aa8   110d
cloud-provider-extraction-migration-pt2   gke-94c9cea96359465dabd0-ce53-3e36-vm_c83c44a7-34fc-45e8-b301-c2eac2972c89   6d1h
cloudbeat-cluster-leader                  cloudbeat-cluster-leader_996k6                                               265d
cluster-autoscaler                        gke-94c9cea96359465dabd0-ce53-3e36-vm                                        390d
elastic-agent-cluster-leader              elastic-agent-leader-elastic-agent-77mkw                                     390d
gcp-controller-manager                    gke-94c9cea96359465dabd0-ce53-3e36-vm                                        297d
gke-common-webhook-lock                   gke-94c9cea96359465dabd0-ce53-3e36-vm_239ec                                  110d
ingress-gce-lock                          gke-94c9cea96359465dabd0-ce53-3e36-vm_d1243                                  209d
kube-controller-manager                   gke-94c9cea96359465dabd0-ce53-3e36-vm_c83c44a7-34fc-45e8-b301-c2eac2972c89   390d
kube-scheduler                            gke-94c9cea96359465dabd0-ce53-3e36-vm_20435fa2-9577-4a48-9d14-66709a20d638   390d
managed-certificate-controller            gke-94c9cea96359465dabd0-ce53-3e36-vm                                        390d
snapshot-controller-leader                gke-94c9cea96359465dabd0-ce53-3e36-vm                                        390d
vpa-recommender                           gke-94c9cea96359465dabd0-ce53-3e36-vm                                        207d

**Comment by Danouchka** [comment]
In my elastic-agent manifest, I have commented the lines below

>  spec:
      # Tolerations are needed to run Elastic Agent on Kubernetes master nodes.
      # Agents running on master nodes collect metrics from the control plane components (scheduler, controller manager) of Kubernetes
      #tolerations:
      #  - key: node-role.kubernetes.io/master
      #    effect: NoSchedule
      #  - key: node-role.kubernetes.io/control-plane
      #    effect: NoSchedule

**Comment by gizas** [comment]
> In my elastic-agent manifest, I have commented the lines below

As long as we speak for cloud k8s instances we dont have access to control-plane so dont think that above will have any effect


> In my kubernetes policy, should I unchecked "leader election" ?

The policy should have "leader election" enabled if we use daemonset deployment

**Comment by Danouchka** [comment]
Hi @gizas 
Node Pool Type changed and all pods moved there. Observing from now to next week 
Will let you know if connection issues happen again

**Comment by vitalyrychkov** [comment]
Have the same error after upgrading 8.7.0 to 8.7.1 on a bare-metal Kubernetes. Had to step back to 8.7.0 because logs were not sent to Elasticsearch

**Comment by Danouchka** [comment]
Hi @gizas so far , so good. 
Since the node pool type change, no incidents !

**Comment by Danouchka** [comment]
Giving more CPUs and Memory on the K8s nodes solved the issue in my opinion 
Let's close the issue

**Comment by Danouchka** [comment]
Hi 

Problem reappears again with a lot of errors like below and 2/3 agents appears offline while stil sending kube-state metrics 
cc @jlind23 @pierrehilbert @gizas 

```
{
  "insertId": "12wpkyd6b5m1wbms",
  "jsonPayload": {
    "@timestamp": "2023-06-04T09:11:02.089Z",
    "message": "Attempting to connect to Elasticsearch version 8.7.1",
    "log.level": "info",
    "service.name": "metricbeat",
    "ecs.version": "1.6.0",
    "component": {
      "binary": "metricbeat",
      "type": "system/metrics",
      "dataset": "elastic_agent.metricbeat",
      "id": "system/metrics-default"
    },
    "log": {
      "source": "system/metrics-default"
    },
    "log.logger": "esclientleg",
    "log.origin": {
      "file.name": "eslegclient/connection.go",
      "file.line": 291
    }
  },
  "resource": {
    "type": "k8s_container",
    "labels": {
      "project_id": "elastic-sa",
      "pod_name": "elastic-agent-4znsv",
      "location": "europe-west1-b",
      "container_name": "elastic-agent",
      "cluster_name": "sa-da-gke-cluster",
      "namespace_name": "kube-system"
    }
  },
  "timestamp": "2023-06-04T09:11:02.096525131Z",
  "severity": "ERROR",
  "labels": {
    "k8s-pod/pod-template-generation": "1",
    "k8s-pod/app": "elastic-agent",
    "compute.googleapis.com/resource_name": "gke-sa-da-gke-cluster-pool-1-99abf297-14fj",
    "k8s-pod/controller-revision-hash": "7b9cff567d"
  },
  "logName": "projects/elastic-sa/logs/stderr",
  "receiveTimestamp": "2023-06-04T09:11:05.686618544Z"
}
```

**Comment by cmacknz** [comment]
It is possible this is the deadlock that in https://github.com/elastic/elastic-agent/issues/2691 that will be fixed in 8.8.1, but we've always seen logs similar to those in https://github.com/elastic/elastic-agent/issues/2628 when this happens.

**Comment by Danouchka** [comment]
Closed with 8.8.1 : more stable
